### DISCUSSION TITLE
best method for creating an audio callback which can take an input stream and an output stream as arguments?

### DISCUSSION THREAD

**Discussion by SnazzyBlanket** [pschatzmann/arduino-audio-tools#1435] [Q&A] [answered]
Sorry if this makes no sense but I am wondering if there is a way to create a Daisy Like AudioCallback which would take input arguments and output arguments (stream in and stream out) and then allow you to apply dsp in between, as well as set flags, start timers, etc. 

In other words a place to test dsp ideas etc without having to create a new stream class- similar to this

static void AudioCallback(AudioHandle::InterleavingInputBuffer  in,
                          AudioHandle::InterleavingOutputBuffer out,
                          size_t                                size)
{
    float output;
    for(size_t i = 0; i < size; i += 2)
    {
        output = noise.Process();

        output = 0.5 * flt.Process(output);

        // left out
        out[i] = output;

        // right out
        out[i + 1] = output;
    }
}

sorry if this seems like an idea that makes no sense

**Comment by pschatzmann** [comment, marked as answer]
I suggest that you have a look at Maximilian: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-maximilian or you might use a [CallbackStream](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_callback_stream.html) to provide or update the audio data.

**Reply by SnazzyBlanket** [reply]
thanks! I will look into it.

**Comment by pschatzmann** [comment]
Today, I have also added the MozziStream class that let's you use [Mozzi](https://sensorium.github.io/Mozzi/).
Here is an [example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-dsp/examples-mozzi/control_gain/control_gain.ino)

**Comment by SnazzyBlanket** [comment]
thanks!! this should be lots of fun
